src/sirocco/workgraph.py

```diff
@@ -224,83 +264,48 @@ def _create_task_node(self, task: core.Task):
             raise NotImplementedError(exc)
 
     def _link_wait_on_to_task(self, task: core.Task):
-        label = AiidaWorkGraph.get_aiida_label_from_graph_item(task)
-        workgraph_task = self._aiida_task_nodes[label]
-        wait_on_tasks = []
-        for wait_on in task.wait_on:
-            wait_on_task_label = AiidaWorkGraph.get_aiida_label_from_graph_item(wait_on)
-            wait_on_tasks.append(self._aiida_task_nodes[wait_on_task_label])
-        workgraph_task.wait = wait_on_tasks
+        self.task_from_core(task).wait = [self.task_from_core(wt) for wt in task.wait_on]
 
     def _link_input_nodes_to_task(self, task: core.Task, input_: core.Data):
         """Links the input to the workgraph task."""
-        task_label = AiidaWorkGraph.get_aiida_label_from_graph_item(task)
-        input_label = AiidaWorkGraph.get_aiida_label_from_graph_item(input_)
-        workgraph_task = self._aiida_task_nodes[task_label]
+        workgraph_task = self.task_from_core(task)
+        input_label = self.get_aiida_label_from_graph_item(input_)
         workgraph_task.add_input("workgraph.any", f"nodes.{input_label}")
 
         # resolve data
-        if (data_node := self._aiida_data_nodes.get(input_label)) is not None:
+        if isinstance(input_, core.AvailableData):
             if not hasattr(workgraph_task.inputs.nodes, f"{input_label}"):
                 msg = f"Socket {input_label!r} was not found in workgraph. Please contact a developer."
                 raise ValueError(msg)
-            socket = getattr(workgraph_task.inputs.nodes, f"{input_label}")
-            socket.value = data_node
-        elif (output_socket := self._aiida_socket_nodes.get(input_label)) is not None:
-            self._workgraph.add_link(output_socket, workgraph_task.inputs[f"nodes.{input_label}"])
+            getattr(workgraph_task.inputs.nodes, f"{input_label}").value = self.data_from_core(input_)
+        elif isinstance(input_, core.GeneratedData):
+            self._workgraph.add_link(self.socket_from_core(input_), workgraph_task.inputs[f"nodes.{input_label}"])
         else:
-            msg = (
-                f"Input data node {input_label!r} was neither found in socket nodes nor in data nodes. The task "
-                f"{task_label!r} must have dependencies on inputs before they are created."
-            )
-            raise ValueError(msg)
+            raise TypeError
```

Comment:
For internal validation with type checker we can actually use `assert_never` https://typing.python.org/en/latest/guides/unreachable.html
```suggestion
            assert_never() # need from typing import assert_never on top
```
one needs to change the signature of the function. I am also okay to do this in a different PR since this was not the goal.

Reply:
Thanks, didn't know about it. Yep, can be another PR.

Reply:
made an issue #128